subscriber.go

```diff
@@ -345,17 +357,37 @@ func (s *Subscriber) Sync(ctx context.Context, peerID peer.ID, nextCid cid.Cid,
 	var err error
 	var syncer Syncer
 
-	// If address specified, then get URL for HTTP sync, or add multiaddr to peerstore.
+	isHttpPeerAddr := false
+	shouldAddToHttpPeerstore := false
 	if peerAddr != nil {
 		for _, p := range peerAddr.Protocols() {
 			if p.Code == multiaddr.P_HTTP || p.Code == multiaddr.P_HTTPS {
-				syncer, err = s.httpSync.NewSyncer(peerID, peerAddr)
-				if err != nil {
-					return cid.Undef, fmt.Errorf("cannot create http sync handler: %w", err)
-				}
+				isHttpPeerAddr = true
+				shouldAddToHttpPeerstore = true
 				break
 			}
 		}
+	}
+
+	if peerAddr == nil {
+		// Check if we have an http url for this peer since we didn't get a peerAddr.
+		// Note that this gives a preference to use httpSync over dtsync if we have
+		// seen http address and we called sync with no explicit peerAddr.
+		possibleAddrs := s.httpPeerstore.Addrs(peerID)
```

Comment:
If we have both addresses, do we want to prefer one over the other?  Or, should we discard the older one?  Or, should we try to sync with one, and if that fails try the other?

If a publisher ever uses an http address, it means they will be stuck using that, even if they later have a libp2p address.  That might be a reason to discard the older address.

Reply:
That has a preference for HTTP for no specific reason. If a publisher ever uses an http address it will keep trying to use that until the addresses' TTL expires (currently 30 minutes).

I think it's okay to not try to use libp2p as a fallback unless we see a strong need for it. But open to discussion.

Reply:
to expand a bit:

1. I think it's unlikely a publisher will have both a libp2p publisher and http publisher for the same peer ID.
2. In the case that they do, it's probably safe to prefer the http publisher since it's less resource intensive for the publisher (an http cache proxy is much easier than a graphsync cache proxy).
3. If the http endpoint fails we eventually remove it from the peerstore via the TTL mechanism.